### PR TITLE
1 add decode test for double, boolean, int, long and null; 2 fix some…

### PR DESCRIPTION
**Pull request by gaoxinge**
- 对hessian2 decode添加了单元测试（double，boolean，int，long，null）
- 根据单元测试，更正了部分decode的行为

### CODE REVIEW

File: decode.go [hunk 1]
```diff
@@ -308,27 +307,26 @@ func (d *Decoder) decInt32(flag int32) (int32, error) {
 	}
 
 	switch {
-	//direct integer
 	case tag >= 0x80 && tag <= 0xbf:
-		return int32(tag - BC_INT_ZERO), nil
+		return int32(int8(tag - BC_INT_ZERO)), nil
```
Comment (AlexStocks) [inline decode.go]: 能否添加注释，说明添加一次转化的意义

File: decode.go [hunk 2]
```diff
@@ -403,34 +401,31 @@ func (d *Decoder) decInt64(flag int32) (int64, error) {
 		return int64(i32), err
 
 	case tag == BC_LONG_INT: // x59
-		i32, err := d.decInt32(TAG_READ)
-		return int64(i32), err
+		var t int32
+		err = binary.Read(d.reader, binary.BigEndian, &t)
```
Comment (AlexStocks) [inline decode.go]: 这个地方的改动，是测试时候测试出来了bug吗？

File: decode.go (hunk 1 above)
Comment (gaoxinge) [reply decode.go]: 当tag小于BC_INT_*时，使用int32直接转还是非负整数，需要使用int8先强转成负整数。对应测试：

```go
func TestInt(t *testing.T) {
	// ...
	testIntFramework(t, "replyInt_m16", -16)
	// ...
}
```

File: decode.go (hunk 2 above)
Comment (gaoxinge) [reply decode.go]: 这是我改了`int32`的`decode`之后，发现下面的测试没通过的。所以我不确定，在不改`int32`的`decode`的情况下，下面的`int64`的`decode`能不能通过。

```go
func TestLong(t *testing.T) {
	// ...
	testLongFramework(t, "replyLong_0x40000", 0x40000)
	// ...
	testLongFramework(t, "replyLong_0x7fffffff", 0x7fffffff)
	// ...
	testLongFramework(t, "replyLong_m0x40001", -0x40001)
	// ...
	testLongFramework(t, "replyLong_m0x80000000", -0x80000000)
	// ...
}
```

File: decode_test.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+	b = append(b, 'z')
+	return b
+}
+
+func sendRequest(method string) []byte {
+	client := &http.Client{}
+
+	req, err := http.NewRequest(
+		"POST",
+		"http://hessian.caucho.com/test/test",
```
Comment (wongoo) [inline decode_test.go]: 这部分是调用远程服务进行测试，依赖于远程服务的状态和版本，如果是能够调用本地的java那是最好。 另外java这部分不了解，我还要看看java代码再做答复。

File: decode.go [hunk 4]
```diff
@@ -308,27 +307,26 @@ func (d *Decoder) decInt32(flag int32) (int32, error) {
...
 		}
-		return int32(tag-BC_INT_BYTE_ZERO)<<8 + int32(buf[0]), nil
+		return int32(int8(tag - BC_INT_BYTE_ZERO)) << 8 + int32(buf[0]), nil
 
 	case tag >= 0xd0 && tag <= 0xd7:
 		if _, err = io.ReadFull(d.reader, buf[:2]); err != nil {
 			return 0, jerrors.Trace(err)
 		}
-		return int32(tag-BC_INT_SHORT_ZERO)<<16 + int32(buf[0])<<8 + int32(buf[1]), nil
+		return int32(int8(tag - BC_INT_SHORT_ZERO)) << 16 + int32(buf[0]) << 8 + int32(buf[1]), nil
```
Comment (wongoo) [inline decode.go]: 此部分，可以由更好的做法，是直接构造byte数组，直接将数组转为相应的类型， 参考做法： https://github.com/vogo/gohessian/blob/master/long.go

File: decode_test.go (hunk 3 above)
Comment (wongoo) [reply decode_test.go]: 我想了一下，你这部分只是针对decode的部分进行了测试，假设服务器返回的是正确内容的字节编码串，那么当前库也encode应该产生同样的内容，这部分是不是也需要测试一下。再保证了和服务器返回内容一致的情况，单元测试是否只需测试encode再decode后的内容和原始一样即可？

Comment (gaoxinge) [reply decode_test.go]: - [官网](http://hessian.caucho.com/)上本来有一个放test api的地方，但是无法访问了。

Hessian Binary Web Service Protocol

后来我在[python-hessian](https://github.com/theatlantic/python-hessian)的[测试](https://github.com/theatlantic/python-hessian/blob/master/tests/test_parser.py)中找到了。所以我就直接把python的测试代码移植了过来。

- 你说的第二个想法很棒！我觉得encode部分可以这样尝试一下。

File: decode.go (hunk 4 above)
Comment (gaoxinge) [reply decode.go]: 问一下 https://github.com/vogo/gohessian/blob/master/long.go#L180 的 `b & 0x08`是不是应该是`b & 0x40`？

Comment (gaoxinge) [reply decode.go]: 问一下 https://github.com/vogo/gohessian/blob/master/long.go#L180 的 `b & 0x08`是不是应该是`b & 0x40`？

Comment (wongoo) [reply decode.go]: @gaoxinge  
1. 这里只是为了判断符号位，其实改为 `0x80` 判断最高位更好, 但三字节范围 `-262144 262143 `其二进制范围表示 `0xfffc0000 0x0003ffff`, 如果是负数，第一个字节范围是`fc-ff`, 所有不管是`0x08`还是`0x40`还是`0x80`都能够判断到；
2. 可以改为 `i8 = int8(u8)` 更简单些；

**Comment by wongoo** [comment]
@gaoxinge  你看上面这些是不是要优化一下再提交？ 我想等这个pr merge后，参考 https://github.com/vogo/gohessian/ 一样调整一下代码结构，方便后面优化。

**Comment by gaoxinge** [comment]
@wongoo 好的，等我先改一下

File: decode.go (hunk 4 above)
Comment (gaoxinge) [reply decode.go]: @wongoo get了，谢谢！

**Comment by gaoxinge** [comment]
@AlexStocks @wongoo 我又改了一版，并添加了对binary和date的测试，你们有空帮我看看有没有什么问题，谢谢！

File: decode.go [hunk 5]
```diff
@@ -806,30 +815,31 @@ func (d *Decoder) decBinary(flag int32) ([]byte, error) {
 		return []byte(""), nil
 	}
 
-	data = make([]byte, 0, CHUNK_SIZE<<1)
-	for tag == BC_BINARY_CHUNK {
+	data := make([]byte, 0)
+	buf := make([]byte, 65536)
+	for {
```
Comment (wongoo) [inline decode.go]: @gaoxinge 这段binary的改造好像有问题，没有看到处理 non-final chunk和 final chunk的逻辑。原来的代码是在处理完 non-final chunk后还有一个读 final chunk的逻辑。

File: decode_test.go (hunk 3 above)
Comment (wongoo) [reply decode_test.go]: @gaoxinge  请求远程服务的这段要改吗？

**Comment by wongoo** [comment]
@gaoxinge 我觉得先不要添加新的修改了，不然PR太大了，怕看不过来，PR越小越单纯越好。

File: decode.go [hunk 6]
```diff
@@ -806,30 +815,31 @@ func (d *Decoder) decBinary(flag int32) ([]byte, error) {
...
 		length, err = d.getBinaryLength(tag)
-		if err != nil || CHUNK_SIZE < length {
-			return nil, jerrors.Annotatef(err, "getBinaryLength(tag:%d) = length:%d", tag, length)
+		if err != nil {
```
Comment (AlexStocks) [inline decode.go]: ** CHUNK_SIZE**是4096，恶意用户可能填一个很长的值，所以此处长度判断不能去掉

File: decode.go [hunk 7]
```diff
@@ -806,30 +815,31 @@ func (d *Decoder) decBinary(flag int32) ([]byte, error) {
 		return []byte(""), nil
 	}
 
-	data = make([]byte, 0, CHUNK_SIZE<<1)
-	for tag == BC_BINARY_CHUNK {
+	data := make([]byte, 0)
```
Comment (AlexStocks) [inline decode.go]: ** CHUNK_SIZE**是4096，恶意用户可能填一个很长的值，所以此处长度判断不能去掉

File: decode.go (hunk 6 above)
Comment (gaoxinge) [reply decode.go]: 是否限制用户填一个很长的值，应该放在应用层去写。个人感觉encode和decode只需要负责根据协议去做序列化和反序列化。

File: decode.go (hunk 5 above)
Comment (gaoxinge) [reply decode.go]: 读binary一共有四种情况，只有当tag为‘A’，需要继续读下一个binary块。按照这个逻辑，我把判断是否需要继续读取的代码放到这里：
https://github.com/dubbogo/hessian2/blob/aa0cac3acb47469c43b256f304c2ecbc1dcfa3ad/decode.go#L833-L840

File: decode_test.go (hunk 3 above)
Comment (gaoxinge) [reply decode_test.go]: 是的，需要把远程服务改成本地版本，另外还需要添加encode的测试代码。不过这次可能来不改，可以先把这个放到项目规划或者issue里面。

File: decode.go (hunk 6 above)
Comment (AlexStocks) [reply decode.go]: ok. agree with u.

File: decode_test.go (hunk 3 above)
Comment (wongoo) [reply decode_test.go]: ok，那先记录一下这个问题。

**Comment by wongoo** [comment]
@gaoxinge @AlexStocks  我没其他问题！ 只需记录下一个问题： `单元测试请求外部服务改为请求本地java代码`.
